tower-http/src/classify.rs

```diff
@@ -0,0 +1,170 @@
+//! Tools for classifying responses as either success or failure.
+
+use http::{HeaderMap, Response, StatusCode};
+use http_body::Body;
+use std::{convert::Infallible, marker::PhantomData};
+
+/// Trait for classifying responses as either success or failure.
+pub trait ClassifyResponse {
+    /// The output of doing the classification.
+    ///
+    /// This could be an HTTP status code or some other kind of status.
+    type Output;
+
+    /// The type used to classify the end of streaming (Eos) responses.
+    type ClassifyEos: ClassifyEos<Output = Self::Output>;
+
+    /// Perform the classification.
+    fn classify_response<B>(
+        &mut self,
+        res: &Response<B>,
+    ) -> ClassifiedNowOrLater<Self::Output, Self::ClassifyEos>
+    where
+        B: Body;
+}
+
+/// Trait for classifying end of streams ("Eos") as either success or failure.
+pub trait ClassifyEos {
+    /// The output of doing the classification.
+    ///
+    /// This could be an HTTP status code or some other kind of status.
+    type Output;
+
+    /// Perform the classification.
+    fn classify_eos(&mut self, trailers: Option<&HeaderMap>) -> Classification<Self::Output>;
+}
+
+/// Result of doing a classification.
+#[derive(Debug)]
+// really don't like this name... :(
```

Comment:
perhaps unsurprisingly i'm not a fan either...we could try to brainstorm something better, but it does get the point across...

Reply:
In linkerd-proxy you write `Produce a stream classifier for this response`. Do you mean a stream like `Stream<Item = _>` which then gets polled somewhere?

Other ideas:

- `TryClassify`
- `ResponseClassification`
- `ClassifiedResponse`

Reply:
"stream" here refers to "streaming bodies", not the `Stream` trait, but yes, it produces a classifier for the response body stream. The name `Stream` might not be good to use here because of conflicts with the trait?

Reply:
I've gone with `ClassifiedResponse` for now.